### DISCUSSION TITLE
how do you make a button appear on screen, which then takes you to another screen?

### DISCUSSION THREAD

**Discussion by DarkiDoki** [FabricMC/fabric-api#2628] [Mod Dev Support] [answered]
how do you make a button appear on screen, which then takes you to another screen? right now I have a "Home" screen that opens with "0" and another screen "Settings". I can't figure out how to add a button **at all**. I want to add a button on the Home that takes you to other screens when you click on their button. I'm using kaupenjoe's tutorial mod as base and cottonmc LibGui for GUI(if there's a way to do it with just Minecraft GUI please tell me I don't know). this is all the code I have besides
        KeyInputHandler.register();
        KeyInputHandler2.register();
inside TutorialModClient
please, how do you add buttons and make them appear on the screen
-----
public class KeyInputHandler {

    public static final String KEY_CATEGORY_TUTORIAL = "key.category.tutorialmod.tutorial";
    public static final String KEY_DRINK_WATER = "key.tutorialmod.drink_water";
    public static KeyBinding drinkingKey;

    public static void registerKeyInputs() {

        ClientTickEvents.END_CLIENT_TICK.register(client -> {
            if(drinkingKey.wasPressed()) {
                // This happens when our custom key is pressed
                MinecraftClient.getInstance().setScreen(new MyScreens(new HomeScreen()));
            }
        });
    }

    public static void register() {
        drinkingKey = KeyBindingHelper.registerKeyBinding(new KeyBinding(
            KEY_DRINK_WATER,
            InputUtil.Type.KEYSYM,
            GLFW.GLFW_KEY_0,
            KEY_CATEGORY_TUTORIAL
        ));

        registerKeyInputs();
    }
}
-----
public class KeyInputHandler2 {
    public static final String KEY_SETTINGS_TUT = "key.category.tutorialmod.tutorial";
    public static final String KEY_OPEN_SETTINGS = "key.tutorialmod.open_settings";
    public static KeyBinding opensettingsKey;

    public static void registerKeyInputs() {
        ClientTickEvents.END_CLIENT_TICK.register(client -> {
            if(opensettingsKey.wasPressed()) {
                // This happens when our custom key is pressed
                MinecraftClient.getInstance().setScreen(new MyScreens(new SettingsScreen()));
            }
        });
    }

    public static void register() {
        opensettingsKey = KeyBindingHelper.registerKeyBinding(new KeyBinding(
                KEY_OPEN_SETTINGS,
                InputUtil.Type.KEYSYM,
                GLFW.GLFW_KEY_K,
                KEY_SETTINGS_TUT
        ));

        registerKeyInputs();
    }
}
-----
public class HomeScreen extends LightweightGuiDescription {
    public HomeScreen(){
        WGridPanel root = new WGridPanel();
        setRootPanel(root);
        root.setSize(300,200);
    }
}
-----
public class SettingsScreen extends LightweightGuiDescription {
    public SettingsScreen(){
        WGridPanel root = new WGridPanel();
        setRootPanel(root);
        root.setSize(300,300);
    }
}
-----
public class MyScreens extends CottonClientScreen {
    public MyScreens(GuiDescription description) {
        super(description);
    }
}

**Comment by warjort** [comment, marked as answer]
I don't know about libgui you would need to talk to whoever writes that mod. 

For a simple example look at vanilla's FatalErrorScreen
i.e. add the button in init() then call super.render() in your render() method to draw it.

**Reply by DarkiDoki** [reply]
add the button in init() then call super.render() in your render() method to draw it.

can you break this down more? i see the init button part and the render but what do I use to make a button appear on the screen I made? i tried copying it and changing FetalErrorScreen to my HomeScreen class but it didn't render on it. Is it because I can't cross-apply the custom GUI with minecraft's GUI?

@Environment(value=EnvType.CLIENT)
public class FatalErrorScreen
extends Screen {
    private final Text message;

    public FatalErrorScreen(Text title, Text message) {
        super(title);
        this.message = message;
    }

    @Override
    protected void init() {
        super.init();
        this.addDrawableChild(new ButtonWidget(this.width / 2 - 100, 140, 200, 20, ScreenTexts.CANCEL, button -> this.client.setScreen(null)));
    }

    @Override
    public void render(MatrixStack matrices, int mouseX, int mouseY, float delta) {
        this.fillGradient(matrices, 0, 0, this.width, this.height, -12574688, -11530224);
        FatalErrorScreen.drawCenteredText(matrices, this.textRenderer, this.title, this.width / 2, 90, 0xFFFFFF);
        FatalErrorScreen.drawCenteredText(matrices, this.textRenderer, this.message, this.width / 2, 110, 0xFFFFFF);
        super.render(matrices, mouseX, mouseY, delta);
    }

    @Override
    public boolean shouldCloseOnEsc() {
        return false;
    }
}

**Reply by warjort** [reply]
Look at the Screen class which FatalErrorScreen extends.

> this.addDrawableChild(new ButtonWidget(this.width / 2 - 100, 140, 200, 20, ScreenTexts.CANCEL, button -> 
> this.client.setScreen(null)));
> 

This adds the button to the list of renderable widgets.

>    super.render(matrices, mouseX, mouseY, delta);

This calls Screen.render() which draws the widgets in that list

As I said above, you need to contact whoever writes your gui library for help with using that.
